NuMPI/Optimization/bugnicourt_cg.py

```diff
@@ -0,0 +1,241 @@
+"""
+Constrained Conjugate Gradient for moderately nonlinear problems.
+This implementation is mainly based upon Bugnicourt et. al. - 2018, OUT_LIN
+algorithm.
+"""
+
+import numpy as np
+import scipy.optimize as optim
+from inspect import signature
+from NuMPI.Tools import Reduction
+
+
+def constrained_conjugate_gradients(fun, hessp,
+                                    x0, mean_val=None,
+                                    gtol=1e-8,
+                                    maxiter=3000,
+                                    callback=None,
+                                    communicator=None,
+                                    bounds=None
+                                    ):
+    '''
+    Constrained conjugate gradient algorithm from Bugnicourt et al. [1].
```

Comment:
Is this something Bugnicourt came up with, or did Bugnicourt only apply this to contact mechanics. I am bit hesitant calling the thing Bugnicourt, as I assume that this is a method someone else came up with that Bugnicourt just adopted to contact mechanics. Are you sure he is the **origin** of this particular CG flavor?

Reply:
What I think we mainly want to say is that our implentation follows their description. 

Vollebregt, E. A. H. J Optim Theory Appl 162, 931–953 (2014) 
(Kalkers code)

came up with the same idea of skipping Polonsky and Keers restart quasi simultaneously (I think Bugnicourt didn't know about this paper. At least they don't cite it).

maybe we should just call it by what it is and cite Bugnicourt and Vollebregt as references.